Code/Tools/ProjectManager/Source/GemCatalog/GemModel.cpp

```diff
@@ -27,53 +31,218 @@ namespace O3DE::ProjectManager
         return m_selectionModel;
     }
 
-    QModelIndex GemModel::AddGem(const GemInfo& gemInfo)
+    void SetItemDataFromGemInfo(QStandardItem* item, const GemInfo& gemInfo, bool metaDataOnly = false)
     {
-        if (FindIndexByNameString(gemInfo.m_name).isValid())
+        item->setFlags(Qt::ItemIsEnabled | Qt::ItemIsSelectable);
+        item->setData(gemInfo.m_name, GemModel::RoleName);
+        item->setData(gemInfo.m_displayName, GemModel::RoleDisplayName);
+        item->setData(gemInfo.m_dependencies, GemModel::RoleDependingGems);
+        item->setData(gemInfo.m_version, GemModel::RoleVersion);
+        item->setData(gemInfo.m_downloadStatus, GemModel::RoleDownloadStatus);
+
+        if (!metaDataOnly)
         {
-            // do not add gems with duplicate names
-            // this can happen by mistake or when a gem repo has a gem with the same name as a local gem
-            AZ_TracePrintf("GemModel", "Ignoring duplicate gem: %s\n", gemInfo.m_name.toUtf8().constData());
-            return QModelIndex();
+            item->setData(false, GemModel::RoleWasPreviouslyAdded);
+            item->setData(gemInfo.m_isAdded, GemModel::RoleIsAdded);
+            item->setData("", GemModel::RoleNewVersion);
         }
+    }
 
-        QStandardItem* item = new QStandardItem();
+    void AddGemInfoVersion(QStandardItem* item, const GemInfo& gemInfo)
+    {
+        QList<QVariant> versionList;
+        auto variant = item->data(GemModel::RoleGemInfoVersions);
+        if (variant.isValid())
+        {
+            versionList = variant.value<QList<QVariant>>();
+        }
+        QVariant gemVariant;
+        gemVariant.setValue(gemInfo);
 
-        item->setFlags(Qt::ItemIsEnabled | Qt::ItemIsSelectable);
+        // sanity check we aren't adding a gem with an existing path and version
+        int versionToReplaceIndex = -1;
+        for (int i = 0; i < versionList.size(); ++i)
+        {
+            const QVariant& existingGemVariant = versionList.at(i);
+            const GemInfo& existingGemInfo = existingGemVariant.value<GemInfo>();
+            if (existingGemInfo.m_path == gemInfo.m_path)
+            {
+                if (existingGemInfo.m_version == gemInfo.m_version)
+                {
+                    AZ_Info("ProjectManager", "Not adding GemInfo because a GemInfo with path (%s) and version (%s) already exists.",
```

Comment:
Nope it does not, it uses Printf

Reply:
It looks like Printf adds a newline:
https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/Debug/Trace.h#L114

Reply:
Actually it doesn't because that ITrace::Printf is overridden by the implementation class of [Trace::Printf](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/Debug/Trace.cpp#L532-L559)